lib/thrift/generator/server/binary_framed.ex

```diff
@@ -0,0 +1,138 @@
+
+    handler = quote do
+      rsp = handler_module.unquote(handler_fn_name)()
+      unquote(build_responder(function.return_type, response_module))
+    end
+    |> wrap_with_try_catch(function, file_group, response_module)
+
+    quote do
+      def handle_thrift(unquote(fn_name), _binary_data, handler_module) do
+        unquote(handler)
+      end
+    end
+  end
+  def generate_handler_function(file_group, service_module, function) do
+    fn_name = Atom.to_string(function.name)
+    args_module = service_module
+    |> Module.concat(Service.service_module_name(function, :args))
+
+    response_module = service_module
+    |> Module.concat(Service.service_module_name(function, :response))
+
+    struct_matches = function.params
+    |> Enum.map(fn param ->
+      {param.name, Macro.var(param.name, nil)}
+    end)
+
+    quote do
+      def handle_thrift(unquote(fn_name), binary_data, handler_module) do
+        case unquote(args_module).BinaryProtocol.deserialize(binary_data) do
+          {%unquote(args_module){unquote_splicing(struct_matches)}, ""} ->
+            unquote(build_handler_call(file_group, function, response_module))
+
+          {_, extra} ->
+            raise %Thrift.TApplicationException{
+              message: "Could not decode #{inspect extra}",
+              type: Thrift.TApplicationException.exception_type(7)}
+        end
+      end
+    end
+  end
+
+  defp build_handler_call(file_group, function, response_module) do
+    handler_fn_name = Utils.underscore(function.name)
+    handler_args = function.params
+    |> Enum.map(&Macro.var(&1.name, nil))
+
+    quote do
+      rsp = handler_module.unquote(handler_fn_name)(unquote_splicing(handler_args))
+      unquote(build_responder(function.return_type, response_module))
+    end
+    |> wrap_with_try_catch(function, file_group, response_module)
+  end
+
+  defp wrap_with_try_catch(quoted_handler, function, file_group, response_module) do
+    rescue_blocks = function.exceptions
+    |> Enum.map(fn exc ->
+      resolved = FileGroup.resolve(file_group, exc)
+
+      dest_module = FileGroup.dest_module(file_group, resolved.type)
+
+      error_var = Macro.var(exc.name, nil)
+      field_setter = quote do: {unquote(exc.name), unquote(error_var)}
+
+      quote do
+        unquote(error_var) in unquote(dest_module) ->
+          serialized_exception = %unquote(response_module){unquote(field_setter)}
+          |> unquote(response_module).BinaryProtocol.serialize()
+          {:reply, serialized_exception}
+      end
+    end)
+
+    catchall = quote do
+      unhandled ->
+        {:server_error, %Thrift.TApplicationException{
+            message: "Server error: #{unhandled.message}",
+            type: Thrift.TApplicationException.exception_type(6)}}
+    end
+
+    rescue_blocks = List.flatten(rescue_blocks ++ [catchall])
+
+    quote do
+      try do
+        unquote(quoted_handler)
+      rescue
+      unquote(rescue_blocks)
```

Comment:
What about throws and exits? Perhaps the catch all should use `catch`? However its possible for functions that raise an "unexpected" error or exit to leak messages or resources. Therefore if there is a catch all to reply on the socket the process would need to exit afterwards. This could be resolved by using a `Task.async` style process but it would be a performance hit. I think it would be preferable to close the socket on these `:server_error`s to prevent these issues without taking a performance hit. How will that affect clients? Is it possible for the server to inform clients that it is shutting down using the thrift protocol?